### PR TITLE
SOLR-9376: [xml] and [json] RawValue DocTransformers should work in cloud mode

### PR DESCRIPTION
**Pull request by magibney**
See: [SOLR-9376](https://issues.apache.org/jira/browse/SOLR-9376)

This PR fixes test coverage to cover both cloud and standalone mode operation of the `[json]` and `[xml]` DocTransformers. Testing could/should definitely be expanded, and there may be some edge cases not accounted for.

(The bulk of the work was actually in updating the test (such as it was) to cover cloud mode. Very tricky because this issue deals with serialization per se, so you need to bypass a bunch of the common test abstractions that purposefully hide serialization details. Wistfully thinking of [SOLR-11872](https://issues.apache.org/jira/browse/SOLR-11872) :slightly_smiling_face:)

There was a false start, reverted. I left the commit history for clarity and comparison, but I think the approach ultimately taken (after the false start) is much better.

### CODE REVIEW

**Comment by magibney** [comment]
[This diff](https://github.com/apache/solr/compare/850b2977c1df9608d434563918e20ebe3168194b..a5ae3464aac01efc658a3b610b9d6e092ecf14fd?expand=1) of the most recent commit cuts down on some of the refactoring noise and gives a better sense of what the essence of the change is

**Comment by epugh** [comment]
Can you explain *why* this didn't work in Cloud mode?  I don't quite see what changed that makes it work in Cloud where it didn't prevously.

**Comment by magibney** [comment]
>Can you explain why this didn't work in Cloud mode? I don't quite see what changed that makes it work in Cloud where it didn't prevously.

As Hoss mentions on this issue, these transformers were applied at the _replica_ level, at which point for cloud `wt=javabin` causes the serialization to (appropriately) not kick in. This issue only affects these "raw value" transformers because they have specifically to do with the serialization, whereas "child doc transformers", e.g. don't depend on serialization so it's not a problem for them to be applied at the replica level.

This PR basically moves the actual "raw value" _actions_ away from the replicas and applies it directly in-line with where (and when) serialization occurs. This I think is appropriate for what these transformers set out to do. I'm happy to go into more detail if it's helpful.

**Comment by epugh** [comment]
Thanks for the explanation, and the use of `wt=javabin` makes sense on the _why_ this is happening.

File: solr/core/src/test/org/apache/solr/response/TestRawTransformer.java [hunk 1]
```diff
@@ -16,55 +16,151 @@
...
+  private static final String DEBUG_LABEL = MethodHandles.lookup().lookupClass().getName();
+
+  /** A basic client for operations at the cloud level, default collection will be set */
+  private static JettySolrRunner JSR;
+  private static HttpSolrClient CLIENT;
 
   @BeforeClass
   public static void beforeClass() throws Exception {
-    initCore("solrconfig-doctransformers.xml", "schema.xml");
+    if (random().nextBoolean()) {
```
Comment (epugh) [inline solr/core/src/test/org/apache/solr/response/TestRawTransformer.java:60]: ONe more thing, is this a common pattenr for doing solrcloud versus NON solr cloud?   It just looks a bit odd to me, is there a better way of swapping the two?

**Review by epugh — CHANGES_REQUESTED** [review]

**Comment by epugh** [comment]
@magibney i just did my first Review in github with some comments...   I'd love to hear your thoughts on my comments, and then I'd love to give it a test and commit!

File: solr/core/src/test/org/apache/solr/response/TestRawTransformer.java (hunk 1 above)
Comment (magibney) [reply solr/core/src/test/org/apache/solr/response/TestRawTransformer.java:60]: Heh; there isn't really a "common pattern" for doing solrcloud vs standalone -- if you can find a better example to pattern after I'm certainly open to alternatives. iiuc this is what [SOLR-11872](https://issues.apache.org/jira/browse/SOLR-11872) sets out to do. This would have been _far_ easier also if it weren't for the fact that we specifically need to exercise the non-javabin serialization formats, and inspect their output directly. So that's _part_ of what made this setup more complex.

File: solr/core/src/test/org/apache/solr/response/TestRawTransformer.java [hunk 2]
```diff
@@ -16,55 +16,151 @@
...
  *
  * See SOLR-7993
  */
-public class TestRawTransformer extends SolrTestCaseJ4 {
+public class TestRawTransformer extends SolrCloudTestCase {
+
+  private static final String DEBUG_LABEL = MethodHandles.lookup().lookupClass().getName();
+
+  /** A basic client for operations at the cloud level, default collection will be set */
+  private static JettySolrRunner JSR;
```
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/test/org/apache/solr/response/TestRawTransformer.java:55]: I take your point. It does try to follow some of the de facto conventions that @hossman seems to have set out in some tests, which are some of the most robust I could find (or at least the best among those with which I'm familiar?) -- e.g. [TestCloudJSONFacetSKGEquiv](https://github.com/apache/solr/blob/c7abf3f32ca439b73fbe0ed5a3a6d70ecc411cbb/solr/core/src/test/org/apache/solr/search/facet/TestCloudJSONFacetSKGEquiv.java#L104).

They're not properly `final` (so all-caps does violate the broader java convention) -- but as used in the test class they're "final-_ish_" ... so idk. I'm fine to change it, though perhaps that change will be obviated by an altogether different approach to cloud vs. standalone?

File: solr/core/src/java/org/apache/solr/response/transform/RawValueTransformerFactory.java [hunk 3]
```diff
@@ -109,19 +131,8 @@ public String getName()
     @Override
     public void transform(SolrDocument doc, int docid) {
       Object val = doc.remove(field);
-      if(val==null) {
```
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/transform/RawValueTransformerFactory.java:112]: Thanks! ... fwiw I wouldn't have characterized this as confusing exactly ... or no more so than the proposed alternative. In any case it's hard to judge from this particular change in isolation. Hopefully overall though this is clearer -- or _differently_ confusing, and in a more appropriate way.

File: solr/core/src/java/org/apache/solr/response/TextResponseWriter.java [hunk 4]
```diff
@@ -55,6 +58,10 @@
 
   protected Calendar cal;  // reusable calendar instance
 
+  private static final ReturnFields DUMMY_RETURN_FIELDS = new SolrReturnFields();
```
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/TextResponseWriter.java:61]: Yeah; `== null` was indeed an obvious first choice; but that doesn't work because it's being used as shorthand to check whether we're in the appropriate state to do special handling. And some of the "inappropriate" cases are signified by `null` explicitly passed, which would have matched on `null`. I could at least comment it or name the variable more clearly. Good point.

File: solr/core/src/java/org/apache/solr/response/RawShimTextResponseWriter.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+public class RawShimTextResponseWriter extends TextResponseWriter {
+
+  private final TextResponseWriter backing;
+
+  RawShimTextResponseWriter(TextResponseWriter backing) {
+    super(null, false);
+    this.backing = backing;
+  }
+
+  // convert non-raw to raw. These are the reason this class exists
```
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/RawShimTextResponseWriter.java:36]: I will add javadoc comments to this class to clarify. This class is needed b/c in the case where control of serialization is delegated to `SchemaField`, the API (appropriately) gives no way to indicate that the serialization should be raw. So this class exists as a shim TextResponseWriter that delegates to another TextResponseWriter, with the slight modification that it causes calls to `writeStr(...)` and `writeArray(...)` to delegate to their "raw" counterparts.

File: solr/core/src/java/org/apache/solr/response/TextResponseWriter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (epugh) [reply solr/core/src/java/org/apache/solr/response/TextResponseWriter.java:61]: Awesome... thanks!

File: solr/core/src/test/org/apache/solr/response/TestRawTransformer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (epugh) [reply solr/core/src/test/org/apache/solr/response/TestRawTransformer.java:55]: Honestly, I think that we might be bikeshedding here, and in lieu of any commonly accepted standard for these types of variable names (that is enforced by a linter!), this is fine.   I looked at the `TestCloudJSONFacetSKGEquiv` and it does use the same pattern.

File: solr/core/src/test/org/apache/solr/cloud/TestRandomFlRTGCloud.java [hunk 6]
```diff
@@ -96,14 +105,12 @@
...
-      // SOLR-9376: RawValueTransformerFactory doesn't work in cloud mode 
-      //
-      // new RawFieldValueValidator("json", "eee_s", "my_json_field_alias"),
-      // new RawFieldValueValidator("json", "fff_s"),
-      // new RawFieldValueValidator("xml", "ggg_s", "my_xml_field_alias"),
-      // new RawFieldValueValidator("xml", "hhh_s"),
-      //
+
+       new RawFieldValueValidator("json", "eee_s", "my_json_field_alias"),
+       new RawFieldValueValidator("json", "fff_s"),
```
Comment (epugh) [inline solr/core/src/test/org/apache/solr/cloud/TestRandomFlRTGCloud.java:110]: Is there a indent that needs removing?  Love we rentables test!

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/XMLResponseParser.java [hunk 7]
```diff
@@ -480,5 +494,41 @@ protected SolrDocument readDocument( XMLStreamReader parser ) throws XMLStreamEx
     }
   }
 
+  /**
+   * Converts raw String input (should valid xml when wrapped in a root element) and converts it to the
```
Comment (epugh) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/XMLResponseParser.java:498]: “Should be”

**Review by epugh — APPROVED** [review]

**Comment by epugh** [comment]
I know this is really a bug fix PR, however if you wanted to make it easier to write new transformers like this, is there any other changes you would make?   This looks great and i can test it on monday

**Comment by magibney** [comment]
>if you wanted to make it easier to write new transformers like this, is there any other changes you would make?

I don't think so. Indeed, if someone went to implement their own custom `DocTransformer` that sought to write directly (unescaped) to output, they would currently replace the field value on each doc with an instance of the `WriteableValue` interface ... and their custom plugin would consequently be incompatible with cloud.

The new way would be to override the new (default no-op) `DocTransformer.getRawFields(Collection<String> addToExisting)` method to add the output field (to be written raw) to the specified collection. Should then be compatible with cloud, but otherwise work the same as currently.

File: solr/core/src/java/org/apache/solr/response/RawShimTextResponseWriter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/RawShimTextResponseWriter.java]: added javadocs to clarify

**Comment by epugh** [comment]
Some great work here!

**Comment by magibney** [comment]
Thanks for the feedback, @epugh! I believe the recent commits have incorporated your suggestions. I also just added an explicit/trivial test expressly for the `[xml]` DocTransformer (in 738b24e6). This test can be run against current `apache-solr/main` to demonstrate that the current behavior of the xml doc transformer is different from the json transformer (and different from what this PR implements/proposes). I'm going to add a comment about this to the jira issue shortly.

**Comment by alessandrobenedetti** [comment]
The pull request appears to be very well done, but it is huge and unfortunately, I don't have a massive experience with this area.
I have referred this PR to my colleague Elia Porciani who has good experience with Transformers, hopefully, this helps!

**Review by dsmiley — APPROVED** [review]
Generally approve.  Just some bits of feedback here and there.  I don't claim to have followed the details in depth but I trust you.

Amazing javadocs!

File: solr/core/src/java/org/apache/solr/response/transform/GeoTransformerFactory.java [hunk 8]
```diff
@@ -144,88 +148,111 @@ public void transform(SolrDocument doc, int docid) throws IOException {
 
     }
     
+    // if source has been renamed, update reference
```
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/transform/GeoTransformerFactory.java:151]: Yes, basically the fact that this class had been doing renaming _outside the context of "simple" renaming that corrected for interactions with other field renames_ was a latent problem (perhaps I could/should add a test to stress this issue -- though I might leave that for another issue?). I'm going to think about this, and the "`FieldRenamer` interface vs. "subclass of `TransformerFactory`" question -- perhaps there's a way to factor some of this logic out so it can live in once place and not be the responsibility of each `FieldRenamer` implementation. I'll report back.

File: solr/core/src/java/org/apache/solr/response/transform/DocTransformer.java [hunk 9]
```diff
@@ -52,6 +53,24 @@ public void setContext( ResultContext context ) {
...
+   * NOTE: normally this will be conditional on the `wt` param in the request, as supplied to the
+   * {@link DocTransformer}'s parent {@link TransformerFactory} at the time of transformer creation.
+   *
+   * @param addToExisting Existing collection to which field names should be added.
+   * @return Collection containing field names to be written raw. If a non-null collection was specified as the
+   * method argument, that same collection should be returned. If the method argument is null, a newly-created
+   * collection should be created and returned if any field renames are necessary (locally created return values
+   * need not be externally modifiable -- i.e., {@link java.util.Collections#singleton(Object)} is acceptable).
+   */
+  public Collection<String> getRawFields(Collection<String> addToExisting) {
```
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/transform/DocTransformer.java:70]: Fair; the way this is usually invoked is directly on a single top level `DocTransformers` instance (extends `DocTransformer`), which collects raw fields from each of its child `DocTransformer` instances ... hence the "collector"/"accumulator" pattern. I think I agree with you though ultimately -- esp. given that most (probably _all_) "leaf" `DocTransformer` instances, if they return anything, will return one field (wrapped as a singleton). Will change this.

File: solr/core/src/java/org/apache/solr/response/transform/TransformerFactory.java [hunk 10]
```diff
@@ -40,6 +41,47 @@ public void init(NamedList<?> args) {
...
+   *
+   * It is harmless for a {@link DocTransformer} that does _not_ in practice rename fields to be returned from a
+   * factory that implements this interface (e.g., for conditional renames?); but doing so opens the possibility of
+   * {@link #create(String, SolrParams, SolrQueryRequest, Map, Set)} being called _after_ fields have been renamed,
+   * so such implementations must still check whether the field with which they are concerned has been renamed ...
+   * and if it _has_, must copy the field back to its original name. This situation also demonstrates the
+   * motivation for separating the creation of {@link DocTransformer}s into two phases: an initial phase involving
+   * no field renames, and a subsequent phase that implement extra logic to properly handle field renames.
+   */
+  public interface FieldRenamer {
```
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/transform/TransformerFactory.java:65]: I considered this (and I guess I'm still considering it). Making this a subclass would clarify the intent, and the expense of some (albeit not particularly useful?) flexibility. I'll think about this in conjunction with the above feedback about the "rename checking" in the `[geo]` transformer.

File: solr/core/src/java/org/apache/solr/response/RawShimTextResponseWriter.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Utility class that delegates to another {@link TextResponseWriter}, but converts normal write requests
+ * into "raw" requests that write field values directly to the delegate {@link TextResponseWriter}'s backing writer.
+ */
+public class RawShimTextResponseWriter extends TextResponseWriter {
```
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/RawShimTextResponseWriter.java:31]: No indeed, will change this to package-private.

**Comment by magibney** [comment]
Thanks for the review, @dsmiley! I'll follow up on these suggestions shortly.

File: solr/core/src/java/org/apache/solr/response/transform/GeoTransformerFactory.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (magibney) [reply solr/core/src/java/org/apache/solr/response/transform/GeoTransformerFactory.java:151]: (I just pushed commits addressing all other feedback but this subclass/interface question -- oh, and did also add tests to stress the renaming issue, as I suggested immediately above)

wrt the subclass/interface question: on further reflection, unfortunately I think it might just make things more convoluted (and less flexible) to try to factor out the "renaming" logic, as opposed to (as currently implemented) delegating that responsibility to each class that implements `FieldRenamer` (at a minimum each DocTransformer needs to parse its own SolrParams to determine "source" field, and the GeoTransformer in some cases _doesn't_ rename a field (instead retrieving source values directly from docValues) -- and once you add hooks for both those cases, it's unclear that anything is really being simplified). 

Unless you feel strongly about it, I'm also inclined to leave this as an interface, because I don't think there's any direct benefit (in terms of factoring out code) to making it a subclass, and although the subclass approach would clarify somewhat the main use of this type, it would also needlessly prevent users from extending existing `TransformerFactory` classes to add support for field renaming. This might be kind of an edge case, but absent a strong preference for strict subclass, I'll probably stick with the flexibility of an interface-based approach.

File: solr/CHANGES-SOLR-9376.txt [hunk 12]
```diff
@@ -0,0 +1,21 @@
+TO BE DROPPED INTO CHANGES.txt (and possibly `solr-upgrade-notes.adoc`?)
+FOR WHATEVER VERSION THIS ULTIMATELY LANDS IN:
+
+Upgrade Notes
```
Comment (dsmiley) [inline solr/CHANGES-SOLR-9376.txt:4]: put in `major-changes-in-solr-9.adoc`; not its own file.

File: solr/core/src/java/org/apache/solr/response/transform/DocTransformers.java [hunk 13]
```diff
@@ -49,6 +50,18 @@ public String getName()
     return str.toString();
   }
 
+  @Override
+  public Collection<String> getRawFields() {
+    Collection<String> fields = new ArrayList<>(size());
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/response/transform/DocTransformers.java:55]: this begs to be implemented using Java streams (including flatten).  Just a thought.

File: solr/core/src/java/org/apache/solr/response/transform/DocTransformer.java [hunk 14]
```diff
@@ -52,6 +53,21 @@ public void setContext( ResultContext context ) {
...
+   * write content directly to output for certain field(s), the names of any such field(s) should be returned
+   *
+   * NOTE: normally this will be conditional on the `wt` param in the request, as supplied to the
+   * {@link DocTransformer}'s parent {@link TransformerFactory} at the time of transformer creation.
+   *
+   * @return Collection containing field names to be written raw, or <code>null</code> if no field names should
+   * be written raw. Any collection returned collection need not be externally modifiable -- i.e.,
+   * {@link java.util.Collections#singleton(Object)} is acceptable.
+   */
+  public Collection<String> getRawFields() {
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/response/transform/DocTransformer.java:67]: There doesn't seem to be a semantic difference in null vs an empty collection, and it would be clearer to return non-null so maybe let's just do that?

File: solr/CHANGES-SOLR-9376.txt (hunk 12 above)
Comment (magibney) [reply solr/CHANGES-SOLR-9376.txt:4]: :+1: Yes, I was drafting it there, but it was never ultimately going to stay in its own file. This is getting close enough now that I'll move it into `major-changes-in-solr-9.adoc`

**Comment by magibney** [comment]
Merged in upstream `main`, and addressed all the latest suggestions -- including from Liftbot, who apparently doesn't like LinkedList :-). I think I fit the note into a reasonable place in `major-changes-in-solr-9.adoc`, but not 100% sure where the right place in that file would be. Thanks again for the review, @dsmiley! ... and please lmk if you notice anything else that could use attention.

**Comment by magibney** [comment]
[Test failure](https://github.com/apache/solr/actions/runs/1726746111/attempts/1) didn't reproduce for me (after removal of the `classMethod` test method specifier): `gradlew :solr:solrj:test --tests "org.apache.solr.client.solrj.impl.CloudHttp2SolrClientBuilderTest.classMethod" -Ptests.jvms=1 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.seed=7B861A6BE1228F59 -Ptests.file.encoding=ISO-8859-1`
Re-ran jobs and tests passed.

**Review by eliaporciani — APPROVED** [review]
Changes look good to me
